### PR TITLE
zkdrop: add ZKBadgeboundERC721 contract

### PR DESCRIPTION
**Pull request by yum0e**

### CODE REVIEW

File: contracts/core/libs/sismo-gated/SismoGatedState.sol [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+      badges = Badges(BADGES_GOERLI_ADDRESS);
+      hydraS1AccountboundAttester = HydraS1AccountboundAttester(
+        HYDRA_S1_ACCOUNTBOUND_GOERLI_ADDRESS
+      );
+    } else if (block.chainid == 80001) {
+      badges = Badges(BADGES_MUMBAI_ADDRESS);
+      hydraS1AccountboundAttester = HydraS1AccountboundAttester(
+        HYDRA_S1_ACCOUNTBOUND_MUMBAI_ADDRESS
+      );
+    } else {
```
Comment (dhadrien) [inline contracts/core/libs/sismo-gated/SismoGatedState.sol:71]: would do:
`else if (block.chainid == 31337 || block.chainid == 1337) { local } else throw`  

(so we sure we don't deploy something weird on a chain

File: contracts/core/libs/sismo-gated/SismoGatedState.sol [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+  mapping(uint256 => bool) private _isNullifierUsed;
+
+  uint256[16] private _storagePlaceHolders;
+
+  /**
+   * @dev Constructor
+   * @param _gatedBadge Badge token id that is required to call the function using `onlyBadgesOwner` modifier
+   */
+  constructor(uint256 _gatedBadge) {
```
Comment (dhadrien) [inline contracts/core/libs/sismo-gated/SismoGatedState.sol:54]: at sismo we use _ for storage

File: contracts/core/libs/sismo-gated/SismoGated.sol [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+  /**************************************
+   * Storage slot
+   * 20 slots
+   * -> 2 slots used
+   * -> 1 used by badges, 1 used by hydraS1AccountboundAttester
+   *
+   * -> 18 free slots
+   **************************************/
+
+  Badges public badges;
```
Comment (leosayous21) [inline contracts/core/libs/sismo-gated/SismoGated.sol:44]: Use immutable here

File: contracts/core/libs/sismo-gated/SismoGated.sol [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    0x069e6B99f4DA543156f66274FC6673442803C587;
+
+  // local contract address
+  address public constant BADGES_LOCAL_ADDRESS = 0xeF5b2Be9a6075a61bCA4384abc375485d5e196c3;
+  address public constant HYDRA_S1_ACCOUNTBOUND_LOCAL_ADDRESS =
+    0xf93A0C43A3466488D416628bf149495285e9f274;
+
+  uint8 public constant MAX_NUMBER_OF_BADGES = 10;
+
+  /**************************************
```
Comment (leosayous21) [inline contracts/core/libs/sismo-gated/SismoGated.sol:35]: Remove the use of storage slot

File: contracts/core/libs/sismo-gated/SismoGated.sol [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+  /**
+   * @dev Check if the user is the owner of the badge, a proof can be supplied to allow non-holders to prove they can mint a badge
+   * @param account Address of the user
+   * @param badgeTokenId Token ID of the badge
+   * @param attester Attester contract used to verify proofs
+   * @param data Bytes containing the user request and the proof associated to it
+   */
+  function _checkBadgeOwnership(
+    address account,
+    uint256 badgeTokenId,
```
Comment (leosayous21) [inline contracts/core/libs/sismo-gated/SismoGated.sol:125]: Implement array directly here + minimalValue

File: contracts/core/libs/sismo-gated/SismoGated.sol (hunk 3 above)
Comment (leosayous21) [reply contracts/core/libs/sismo-gated/SismoGated.sol:44]: In fact it's not possible, `Cannot write to immutable here: Immutable variables cannot be initialized inside an if statement`

File: contracts/core/libs/sismo-gated/SismoGated.sol [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+   * @param gatedBadgeTokenId Token ID of the badge that the user needs to hold to access the function
+   * @param gatedBadgeMinimumValue Minimum value of the badge that the user needs to hold to access the function
+   * @param data Data containing the user request and the proof associated to it
+   * @param attester Attester contract used to verify the proof
+   * @notice a proof can also be sent to allow non-holders to prove their eligibility
+   */
+  modifier onlyBadgeOwner(
+    address badgeOwnerAddress,
+    uint256 gatedBadgeTokenId,
+    uint256 gatedBadgeMinimumValue,
```
Comment (dhadrien) [inline contracts/core/libs/sismo-gated/SismoGated.sol:63]: balanceMin

=> will need to add in comments that balance == level for a badge

File: contracts/core/libs/sismo-gated/SismoGated.sol [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+    uint256[] memory gatedBadgeMinimumValues = new uint256[](1);
+    gatedBadgeMinimumValues[0] = gatedBadgeMinimumValue;
+
+    Attester[] memory attesters = new Attester[](1);
+    attesters[0] = attester;
+
+    bytes[] memory dataArray = new bytes[](1);
+    dataArray[0] = data;
+
+    checkBadgesOwnership(
```
Comment (dhadrien) [inline contracts/core/libs/sismo-gated/SismoGated.sol:79]: does not do what it says it does

File: contracts/tests/MockGatedERC721.sol [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+   */
+  function isNullifierUsed(uint256 nullifier) public view returns (bool) {
+    return _isNullifierUsed[nullifier];
+  }
+
+  /**
+   * @dev Marks a nullifier as used
+   * @param nullifier Nullifier to mark as used
+   */
+  function _markNullifierAsUsed(uint256 nullifier) internal {
```
Comment (dhadrien) [inline contracts/tests/MockGatedERC721.sol:100]: _setNullifierAsUsed

File: contracts/core/libs/sismo-gated/SismoGated.sol [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+
+  /**
+   * @dev Check if the user is the owner of the badge, a proof can be supplied to allow non-holders to prove they can mint a badge
+   * @param account Address of the user
+   * @param badgeTokenIds Token ID of the badges
+   * @param badgeMinimumValues Minimum value of the badges
+   * @param attesters Attester contracts used to verify proofs
+   * @param dataArray Array of bytes containing the user requests and the proofs of each badge eligibility
+   */
+  function checkBadgesOwnership(
```
Comment (dhadrien) [inline contracts/core/libs/sismo-gated/SismoGated.sol:98]: does not do what it says it does

File: contracts/core/libs/sismo-gated/SismoGated.sol [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+  /**
+   * @dev Modifier allowing only the owners of the `GATED_BADGE` to trigger the function
+   * @param badgeOwnerAddress Address of the badge owner
+   * @param gatedBadgeTokenId Token ID of the badge that the user needs to hold to access the function
+   * @param gatedBadgeMinimumValue Minimum value of the badge that the user needs to hold to access the function
+   * @param data Data containing the user request and the proof associated to it
+   * @param attester Attester contract used to verify the proof
+   * @notice a proof can also be sent to allow non-holders to prove their eligibility
+   */
+  modifier onlyBadgeOwner(
```
Comment (dhadrien) [inline contracts/core/libs/sismo-gated/SismoGated.sol:60]: does not do what it says it does

File: contracts/core/libs/sismo-gated/SismoGated.sol [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+    Attestation[] memory attestations = attester.generateAttestations(request, proofData);
+
+    return attestations[0];
+  }
+
+  /**
+   * @dev Build the request and the proof from data
+   * @param data Bytes containing the user request and the proof associated to it
+   */
+  function _buildRequestAndProof(
```
Comment (dhadrien) [inline contracts/core/libs/sismo-gated/SismoGated.sol:136]: maybe _decodeRequestAndProofFromBytes

File: contracts/tests/MockGatedERC721.sol [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    _mint(to, tokenId);
+  }
+
+  function safeTransferFrom(
+    address from,
+    address to,
+    uint256 tokenId,
+    bytes memory data
```
Comment (dhadrien) [inline contracts/tests/MockGatedERC721.sol:68]: rename data to sismoProofData 

This is the hard part, they must understand that this is what they will receive via the off-chain prove with sismo flow

File: contracts/tests/MockGatedERC721.sol [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+  ) public override(ERC721) {
+    proveWithSismo(HYDRA_S1_ACCOUNTBOUND_ATTESTER, data);
```
Comment (dhadrien) [inline contracts/tests/MockGatedERC721.sol:70]: I think we should check if attestation destination == to

Comment (dhadrien) [reply contracts/tests/MockGatedERC721.sol:70]: should also check whether attestation value >= badgeLevelMin

**Review by dhadrien — CHANGES_REQUESTED** [review]

File: contracts/core/libs/sismo-gated/SismoGated.sol [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+
+  /**
+   * @dev Check if the `account` address holds badges or is used in valid sismo proofs
+   * @param account Address which holds badges or is used in valid sismo proofs
+   * @param badgeTokenIds Token ID of the badges
+   * @param minBalances Minimum balances (= levels) of the badges
+   * @param attesters Attester contracts used to verify proofs
+   * @param sismoProofDataArray Array of bytes containing the user requests and the proofs of each badge eligibility
+   */
+  function checkAccountBadgesOrSismoProofs(
```
Comment (dhadrien) [inline contracts/core/libs/sismo-gated/SismoGated.sol:100]: I think it still does not check min value with proof with sismo.

I can generate an attestation of level 0

File: contracts/core/libs/sismo-gated/SismoGated.sol [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+   * @param badgeTokenIds Token IDs of the badges
+   * @param minBalances Minimum balances (= levels) of the badges
+   * @param isInclusive If true, the user must hold at least one badge with the minimum balance (=level) inputed.
+   *                    If false, it must hold them all with the minimum balances (=level) inputed.
+   */
+  function checkIfAccountHoldsBadgesWithRequiredLevels(
+    address account,
+    uint256[] memory badgeTokenIds,
+    uint256[] memory minBalances,
+    bool isInclusive
```
Comment (leosayous21) [inline contracts/core/libs/sismo-gated/SismoGated.sol:88]: Very hard to know that it corresponds to an AND or OR operation i think

File: contracts/tests/MockGatedERC721.sol [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+
+    uint256[] memory badgeMinimumValues = new uint256[](2);
+    badgeMinimumValues[0] = GATED_BADGE_MIN_LEVEL;
+    badgeMinimumValues[1] = GATED_BADGE_MIN_LEVEL;
+
+    bool isInclusive = true;
+
+    checkIfAccountHoldsBadgesWithRequiredLevels(to, badgeTokenIds, badgeMinimumValues, isInclusive);
+
+    uint256 nullifier = _getNulliferForAddress(to);
```
Comment (leosayous21) [inline contracts/tests/MockGatedERC721.sol:101]: How you can have only one nullifier but 2 badges? 
GATED_BADGE_TOKEN_ID is used directly inside the `_getNulliferForAddress` function here

File: contracts/tests/MockGatedERC721.sol [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+   * @param from Address to transfer the NFT from
+   * @param to Address to transfer the NFT to
+   * @param tokenId Token ID of the NFT
+   */
+  function safeTransferFrom(
+    address from,
+    address to,
+    uint256 tokenId
+  ) public override(ERC721) ERC1155Gated(to, GATED_BADGE_TOKEN_ID, GATED_BADGE_MIN_LEVEL) {
+    _transfer(from, to, tokenId);
```
Comment (leosayous21) [inline contracts/tests/MockGatedERC721.sol:121]: This function can't work as you could send your NFT to anyone

File: contracts/tests/MockGatedERC721.sol [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+      revert AccountAndRequestDestinationDoNotMatch(to, request.destination);
+    }
+
+    (, uint256[] memory values) = proveWithSismo(
+      HYDRA_S1_ACCOUNTBOUND_ATTESTER,
+      request,
+      sismoProof
+    );
+
+    for (uint256 i = 0; i < values.length; i++) {
```
Comment (leosayous21) [inline contracts/tests/MockGatedERC721.sol:149]: You don't check the different tokenID here

File: contracts/tests/MockGatedERC721.sol [hunk 19]
```diff
@@ -0,0 +1,191 @@
...
+      if (values[i] < GATED_BADGE_MIN_LEVEL) {
+        revert BadgeLevelIsLowerThanMinBalance(values[i], GATED_BADGE_MIN_LEVEL);
+      }
+    }
+
+    safeTransferFrom(from, to, tokenId);
```
Comment (leosayous21) [inline contracts/tests/MockGatedERC721.sol:155]: We don't check the nullifier here?

**Comment by Crocozyabro** [comment]
review tests👍

File: contracts/tests/zk-drop/ZKBadgeboundERC721.sol [hunk 20]
```diff
@@ -18,8 +19,8 @@ import {Pausable} from '@openzeppelin/contracts/security/Pausable.sol';
  * Each time a NFT is minted or transferred, we make sure that the ZK Badge owned by the account is the same as the one used to mint or transfer the NFT
  */
 
-contract ZKBadgeboundERC721 is ERC721, UsingSismo, AccessControl, Pausable {
-  uint256 public constant MERGOOOR_PASS_BADGE_TOKEN_ID = 200002;
+contract ZKBadgeboundERC721 is ERC721Upgradeable, UsingSismo, AccessControl, Pausable {
+  uint256 public immutable MERGOOOR_PASS_BADGE_TOKEN_ID;
```
Comment (dhadrien) [inline contracts/tests/zk-drop/ZKBadgeboundERC721.sol:23]: rename to GATING_BADGE_TOKEN_ID

File: contracts/tests/zk-drop/ZKBadgeboundERC721.sol [hunk 21]
```diff
@@ -32,12 +33,22 @@ contract ZKBadgeboundERC721 is ERC721, UsingSismo, AccessControl, Pausable {
   error NoTokenTransferWhilePaused();
 
   constructor(
+    string memory name,
+    string memory symbol,
+    string memory baseTokenURI,
+    uint256 passTokenId
```
Comment (dhadrien) [inline contracts/tests/zk-drop/ZKBadgeboundERC721.sol:39]: rename

File: tasks/deploy-tasks/tests/deploy-zk-badgebound-erc721.task.ts [hunk 22]
```diff
@@ -54,4 +59,5 @@ task('deploy-zk-badgebound-erc721')
   .addParam('name', 'Name of the token')
   .addParam('symbol', 'Symbol of the token')
   .addParam('tokenURI', 'Token URI')
+  .addParam('passTokenId', 'Token ID of the pass token')
```
Comment (dhadrien) [inline tasks/deploy-tasks/tests/deploy-zk-badgebound-erc721.task.ts:62]: rename param

**Pull request merged**